keps/sig-scheduling/5075-dra-consumable-capacity/README.md

```diff
@@ -0,0 +1,1256 @@
+// ResourceRequest is a per-device resource request specification.
+type ResourceRequest struct {
+	// All marks requesting all resource from the shared device.
+	// If All is set to true, Quantity will be ignored.
+	//
+	// +optional
+	// +default=false
+	All bool `json:"all" protobuf:"bytes,1,name=all"`
+
+	// Requests define a set of requested per-device amount of each resource.
+	// This field is ignored if All is true.
+	// nil set implies requesting all resources.
+	// A zero-length set implies an infinity consumable capacity and requesting zero resource.
+	//
+	// +optional
+	Requests map[QualifiedName]resource.Quantity `json:"requests,omitempty" protobuf:"bytes,2,rep,name=requests"`
+}
+```
+
+#### ResourceClaimStatus's DeviceRequestAllocationResult
+
+```go
+type DeviceRequestAllocationResult struct {
+  ...
+	// AllocatedShare indicates a per-device resource amount allocated by the claim request.
+	// Only the capacity defined in consumableCapacity can be partially allocated.
+	// A summation of allocated resource must be less than or equal each corresponding capacity.
+	// This field lists all consumableCapacity of the device.
+	// nil if the device has no consumableCapacity defined (not sharable).
+	//
+	// +optional
+	// +featureGate=ConsumableCapacity
+	AllocatedShare *map[QualifiedName]resource.Quantity `json:"allocatedShare" protobuf:"bytes,6,opt,name=allocatedShare"`
+}
+```
+
+### Examples
+#### ResourceSlice
+- ResourceSlice with a consumable capacity (A type)
+  ```yaml
+  apiVersion: resource.k8s.io/v1beta1
+  kind: ResourceSlice
+  ...
+  spec:
+    driver: cni.dra.networking.x-k8s.io
+    devices:
+    - name: eth1
+      basic:
+        attributes:
+          name:
+            string: "eth1"
+        consumableCapacity:
+          bandwidth:
+            value: "10Gi"
+  ```
+- ResourceSlice with an infinite consumable capacity (B type)
+  ```yaml
+  apiVersion: resource.k8s.io/v1beta1
+  kind: ResourceSlice
+  ...
+  spec:
+    driver: cni.dra.networking.x-k8s.io
+    devices:
+    - name: eth1
+      basic:
+        attributes:
+          name:
+            string: "eth1"
+        consumableCapacity:
+          ipvlan:
+            infinity: true
+          macvlan:
+            infinity: true
+  ```
+- Combination of both types are also possible.
+#### ResourceClaim
+- ResourceClaim with a resource request (A-type ResourceSlice)
+  ```yaml
+  apiVersion: resource.k8s.io/v1beta1
+  kind: ResourceClaim
+  ...
+  spec:
+    devices:
+      requests:
+      - name: net1
+        deviceClassName: vlan-cni.networking.x-k8s.io
+        resources:
+          requests:
+            bandwidth: "1Gi"
+  ...
+  status:
+    reserveFor:
+    - name: example-pod
+      ...
+    allocation:
+      devices:
+        results:
+        - request: net1
+          device: eth1
+          allocatedShare:
+            bandwidth: "1Gi"
+  ```
+- ResourceClaim without a resource request (request full available device) (A-type ResourceSlice)
```

Comment:
we can already do this, no? why do we need to create this exception in this model, so we have two ways of doing the same